sql-parser/src/main/java/io/crate/sql/Literals.java

```diff
@@ -24,14 +24,163 @@
+                        throw new IllegalArgumentException(ESCAPED_UNICODE_ERROR);
+                    }
+                    builder.append(numericCharArrayToCharArray(tokenCharArray, 1, tokenIdx, 16));
+                    break;
+                case 'x':
+                    // handle hex byte case - up to 2 chars for hex value
+                    tokenIdx = lastMatchedCharIndex(tokenCharArray, 1, 2, Literals::isHexDigit);
+                    if (tokenIdx > 1) {
+                        builder.append(numericCharArrayToCharArray(tokenCharArray, 1, tokenIdx, 16));
+                    } else {
+                        // hex sequence unmatched - output original input
+                        tokenIdx = 0;
+                    }
+                    break;
+                case '0':
+                case '1':
+                case '2':
+                case '3':
+                    // handle octal case - up to 3 chars in total (first one is already "consumed" though)
+                    tokenIdx = lastMatchedCharIndex(tokenCharArray, 1, 2, Literals::isOctalDigit);
+                    builder.append(numericCharArrayToCharArray(tokenCharArray, 0, tokenIdx, 8));
+                    break;
+                default:
+                    // non valid escaped char sequence - output original input
+                    builder.append('\\');
+                    tokenIdx = 0;
+            }
+            // append the rest of token
+            builder.append(tokenCharArray, tokenIdx, tokenLen - tokenIdx);
+        }
+        return builder.toString();
+    }
+
+    private static boolean isOctalDigit(final char ch) {
+        return ch >= '0' && ch <= '7';
+    }
+
+    private static boolean isHexDigit(final char ch) {
+        return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
+    }
+
+    private static char[] numericCharArrayToCharArray(char[] numerical,
+                                                      int beginIndex,
+                                                      int endIndex,
+                                                      int radix) {
+        return Character.toChars(parseInt(numerical, beginIndex, endIndex, radix));
+    }
+
+    /**
+     * Returns the index of the last character of a char array
+     * (within [beginIndex, endIndex) range)
+     * matching a given predicate
+     */
+    private static int lastMatchedCharIndex(char[] seq,
+                                            int beginIndex,
+                                            int endIndex,
+                                            Predicate<Character> predicate) {
+        int idx = beginIndex;
+        final int end = Math.min(endIndex, seq.length - beginIndex);
+        while (idx <= end && predicate.test(seq[idx])) {
+            idx++;
+        }
+        return idx;
+    }
+
+    private static int parseInt(char[] s, int beginIndex, int endIndex, int radix) throws NumberFormatException {
```

Comment:
added this shorten version of `Integer::parseInt` as I could not find one in standard lib operating on `char[]`. Is there an alternative that you know of?